QHyper/problems/Knapsack.py

```diff
@@ -0,0 +1,131 @@
+import random
+from collections import namedtuple
+
+from .problem import Problem
+
+ 
+Item = namedtuple('Item', "weight value")
+
+
+class Knapsack:
+    """Knapsack class"""
+
+    def __init__(self, max_weight: int, N: int=0, max_item_value: int=10) -> None:
+        self.items: list[Item] = []
+        self.all_items: int = N
+        self.max_weight: int = max_weight
+        self.max_item_value: int = max_item_value
+        self.generate_knapsack(N)
+
+    def generate_knapsack(self, N: int) -> None:
+        for _ in range(N):
+            self.items.append(Item(
+                    random.randint(1, self.max_weight), 
+                    random.randint(1, self.max_item_value)
+                )
+            )
+
+    def set_knapsack(self, items: list[tuple[int, int]]) -> None:
+        self.items = [Item(weight, value) for weight, value in items]
+        self.all_items = len(items)
+
+
+class KnapsackProblem(Problem):
+    """Class defining objective function and constraints for knapsack problem
+    
+    Attributes
+    ----------
+    objective_function : str
+        objective function in SymPy syntax
+    constraints : list[str]
+        list of constraints in SymPy syntax
+    wires : int
+        number of qubits in the circuit, equals to number of items in knapsack + max weight of knapsack
+    """
+
+    def __init__(
+        self, 
+        knapsack: Knapsack, 
+        # number_of_layers: int = 6, 
+        # optimization_steps: int = 70,
+        # optimizer: qml.GradientDescentOptimizer = None
+        # optimizer: qml.GradientDescentOptimizer = qml.AdamOptimizer(
+        #     stepsize=0.01,
+        #     beta1=0.9,
+        #     beta2=0.99
+        # )
+    ) -> None:
+        """
+        Parameters
+        ----------
+        knapsack : Knapsack
+            knapsack object with available items and max weight
+        """
+        # self.number_of_layers = number_of_layers
+        # self.optimization_steps = optimization_steps
+        # self.optimizer = optimizer
+        self.wires = knapsack.all_items + knapsack.max_weight
+        self.knapsack = knapsack
+        # self.dev = qml.device("default.qubit", wires=self.wires)
+        self._create_objective_function(knapsack)
+        self._create_constraints(knapsack)
+
+    # def _x(self, wire):
+    #     return qml.Hamiltonian([0.5, -0.5], [qml.Identity(wire), qml.PauliZ(wire)])
+
+    def _create_objective_function(self, knapsack: Knapsack) -> None:
+        xs = [f"x{i}" for i in range(knapsack.all_items)]
+        equation = "-("
+        for i, x in enumerate(xs):
+            equation += f"+{knapsack.items[i].value}*{x}"
```

Comment:
We'll probably soon change this function, so for now it can be left as it is. Idk if this will be helpful in this case, but we should take into consideration that the string can be long, and maybe concatenating it using some other way e.g., with ".join()" might be more efficient.